### ISSUE TITLE
Unable to add auth headers to openseadragon requests

### ISSUE THREAD

**Issue by paulmharwood** [openseadragon/openseadragon#1388]
Hi,

I'm sure I'm missing something fundamental here as no-one else is mentioning issues with adding headers, but I'll risk the shame and ask for help!

In our project we had openseadragon working fine, however I added security (JWT) to a backend where the images are stored, now we need all requests from openseadragon to contain the token in the http header.  I've used the sample code from this project located in the tests for adding headers:

``` javascript
var viewer = OpenSeadragon({
				id: viewerElementID,
			        prefixUrl: "./assets/scripts/openseadragon-bin-2.3.1/images/",
			        zoomInButton:   "zoomin",
			        zoomOutButton:  "zoom-out",
			        homeButton:     "home",
			        fullPageButton: "full-page",
			        showNavigator:  true,
			        navigatorId:   navigatorDivID,
			    	toolbar:       toolbarDivID,
			    	showNavigator: true,
		                loadTilesWithAjax: true,
		                ajaxHeaders: {
		                    "Authorization": localStorage.getItem('token')
		            }
	    	  });
		
		return viewer;
```

Setting the tile source is as per the simple example from the website:

```javascript
$scope.viewer.open([{
	        	type: 'image',
	            url: $scope.selectedImageSrc
	            
	        }]);
```

But in the dev tools in Chrome I can't see any headers being added to the image http requests (403 - as expected).

Am I able to get some pointers on how this should be done correctly - I would appreciate
any help!

Thanks.

**Comment by iangilman** [comment]
It appears that the simple image tile source doesn't support `loadTilesWithAjax`. Thanks for finding this! @seanich, @avandecreme, any interest from either of you to implement that?

@paulmharwood  are you using the simple image tile source just for testing or is that what you need for the real thing?If you're just testing with it, try switching to a different tile source such as DZI. If it's what you need for real, I'm happy to give you pointers on how to fix this issue (if neither Sean nor Antoine are up for it).

**Comment by seanich** [comment]
I can look into this over the weekend :)

**Comment by paulmharwood** [comment]
Thanks for the quick responses both.  

Thanks @seanich for offering to fix, I'm new to Angular and openseadragon but if I need to get involved I'm happy to give it a go.

@iangilman - we're using the simple tile source for the real thing, that's because it worked and we don't really know any better to be honest.  But I'll take a look at DZI and the other examples for other approaches as a plan b.

Cheers!!

**Comment by seanich** [comment]
Okay, I've found the issue here. It turns out `ImageTileSource` is unique in that it loads all its image data in its implementation of `TileSource.getImageInfo` (see [here](https://github.com/openseadragon/openseadragon/blob/master/src/imagetilesource.js#L105-L144)). It's necessary to load the image here to retrieve its height and width, but the implementation currently bypasses `TiledImage.loadTile` which means that options like `loadTilesWithAjax` and `ajaxHeaders` will be ignored. Obviously this is a problem when loading the image requires an `Authentication` header. :)

I'll look into how to modify the implementation of `ImageTileSource` to respect the image loading options specified in the viewer.

**Comment by iangilman** [comment]
@paulmharwood How big are your images (in pixels)? If they are only a few thousand pixels wide, there's not need to do something more intense than the simple tile source (assuming we can get this issue sorted out).

@seanich Yes, I think that's exactly the issue. Maybe we need some sort of general image loading function that both places can call? Btw, @avandecreme wrote the `ImageTileSource`, if you need to ask questions... Thank you for digging into it!

**Comment by paulmharwood** [comment]
@iangilman nothing more than a few thousand pixels wide, simple tile source seems like the right approach on paper - pending the fix of course.  Is there any other approach that can be manipulated just to take a url (with headers) or is the only way with simple tile source and a fix?  Thanks.

**Comment by iangilman** [comment]
Yes, sounds like the right solution. That said, you could also use the legacy tile source with just a single tile: 

http://openseadragon.github.io/examples/tilesource-legacy/

That might be a decent work around for the time being.

**Comment by iangilman** [comment]
@enekonieto It doesn't look like anyone has fixed this. If you're interested in doing so, I'm happy to point you in the right direction and answer questions :)

**Comment by enekonieto** [comment]
In the last years I'm used to program in typescript, all the stuff making "objects" with functions is quite obscure for me. It is a pity but we take a workaround to avoid the bug/missing feature.

Thank you @iangilman for the help offer, I really considered to try fixing it.

**Comment by hrghauri** [comment]
I tried it and I can see header being added. However I do see 2 requests being made, one with the original params and headers and another whose url is `blob:{baseurl}/{someguid}`

**Comment by iangilman** [comment]
@hrghauri Strange! This is with simple image tile source? Can you share your repro?

**Comment by hrghauri** [comment]
@iangilman  Oh no,  I was not actually using simple image tile source. I have not tried with simple image tile source yet.

**Comment by iangilman** [comment]
I see. Even though the title of this issue isn't about simple image, the original post at the top is about simple image, and that's what most of the discussion has been about. It's probably worth starting a new issue if we also have a problem with off headers on custom tile sources. I'll update the title of this issue.

**Comment by hrghauri** [comment]
@iangilman my apologies for diverting the issue. In any case, I do not have any issue with custom tile source so far. Cheers!

**Comment by alimoeeny** [comment]
@iangilman can you please point me to where I can change the code (and submit a PR if successful) so that even a simple image or a dzi url as a tile source can also use the ajaxHeaders? 
I essentially want all the calls to use the auth headers I provide.

**Comment by iangilman** [comment]
@alimoeeny Great! Probably good to start with simple image and branch out from there as needed. Here's where the image is getting loaded: 

https://github.com/openseadragon/openseadragon/blob/640526b444536dccb2c78eec7a0b4b7eeb1eb43c/src/imagetilesource.js#L143

As you can see, it's just using a browser image to load it. As far as I know, you can't add arbitrary headers to a browser image load, so we'll have to add the ability to load it via ajax. Here's where other tiles are loaded via ajax:

https://github.com/openseadragon/openseadragon/blob/640526b444536dccb2c78eec7a0b4b7eeb1eb43c/src/tilesource.js#L780

I wonder if that function (`TileSource.downloadTileStart`) could be called directly from `ImageTileSource`. If not, it would be good to refactor it so we have a common function that can be called from both places, so we don't have to duplicate all that code.

Of course the other part of making this into a PR is making sure that `loadTilesWithAjax` and `ajaxHeaders` are properly piped through.

Sound fair?

**Comment by alimoeeny** [comment]
Thank you @iangilman 
Very helpful info. I'll read through the code and see how far I can go. 
I am sure I will be reaching out to you with questions. 
My guess (maybe naively) is that the harder part is passing the headers through as you mentioned. I hope the loading part can be done elegantly and more importantly without a performance penalty. 
I'll keep you posted.

**Comment by iangilman** [comment]
Fantastic!

Passing the headers is fairly well-established... Search for `loadTilesWithAjax` to see how it's done elsewhere.

Keep me posted 😊

**Comment by DennisFriedl** [comment]
> Thank you @iangilman Very helpful info. I'll read through the code and see how far I can go. I am sure I will be reaching out to you with questions. My guess (maybe naively) is that the harder part is passing the headers through as you mentioned. I hope the loading part can be done elegantly and more importantly without a performance penalty. I'll keep you posted.

Did you have any success? We are facing the same issue at the moment and are wondering what the state of affairs is and where we could help.

**Comment by alimoeeny** [comment]
Unfortunately I have been very busy with other things and I have not yet managed to spend time on this. I have not abandoned it, but if you have time and want to take it on please go ahead.

**Comment by The-Aurora** [comment]
So I have come across this thread as I was also encountering an issue where authorization headers aren't added to my requests.

What I found out is that for tiles, they would have the header present with OPTIONS requests preceeding them (getting the required CORS options).

However, the INITIAL request getting the .dzi file would not. I have temporarily fixed this on the BE by allowing requests ending with .dzi without authorization.

`
        OnPrepareResponse = (ctx) =>
        {
            var context = ctx.Context;
            if (!context.Request.Path.Value!.EndsWith(".dzi", StringComparison.OrdinalIgnoreCase) &&
                !context.User.IsAuthenticated())
            {
                context.Response.StatusCode = 403;
                context.Response.ContentLength = 0;
                context.Response.Body = Stream.Null;
            }
            return;
        },
`

**Comment by The-Aurora** [comment]
> [@The-Aurora](https://github.com/The-Aurora) Thanks for the info on changing the backend!
> 
> This issue is for the simple image tile source, not DZI. I thought the DZI file could already be loaded with auth headers. Are you including `ajaxWithCredentials` and `ajaxHeaders`? What are you passing?
> 
> Also, since this is a different topic than the main issue here, we should probably start a new issue.

Hi,

Sorry, I am short on time and I don't have the energy to create a well-written issue so best I can do right now is dump some bits

The image is built on `node:20.8.1` and I have tried using the latest `"openseadragon": "5.0.1"`

I am configuring OSD like so and verified that the behaviour is still the same, authorization headers on tiles but not on the initial .dzi call.

Might have something to do with CORS or the Options request not being fired prior to the .dzi call. Don't know. Turning off authorization for the .dzi call made it so I can work with it.

```
  private initializeOpenSeadragon(): void {
    const container = this.osdContainer!.nativeElement;

    this.viewer = new OpenSeadragon.Viewer({
      element: container,
      // debugMode: true,
      showNavigationControl: false,

      showNavigator: true,
      navigatorId: 'navigatorDiv',
      navigatorDisplayRegionColor: '#f9df57',

      visibilityRatio: 1,
      maxZoomPixelRatio: 20,

      preserveImageSizeOnResize: true,
      blendTime: 0.1,
      animationTime: 0.5,
      gestureSettingsMouse: {
        clickToZoom: false,
      },

      ajaxHeaders: {
        Authorization: this.apiRequestConfiguration.getToken!,
      },
      ajaxWithCredentials: true,
      loadTilesWithAjax: true,

      tileSources:
        environment.apiUrl +
        '/wsi/' +
        this.sampleImage?.path.replace(/^\/media\/wsi\//, ''),
    });

    // This is necessary, because Types of OSD actually doesn't have navigator.loadTilesWithAjax parameter
    // @ts-expect-error-error
    this.viewer.navigator.ajaxWithCredentials = true;
    // This is necessary, because Types of OSD actually doesn't have navigator.loadTilesWithAjax parameter
    // @ts-expect-error-error
    this.viewer.navigator.loadTilesWithAjax = true;
    // This is necessary, because Types of OSD actually doesn't have setAjaxHeaders method
    // @ts-expect-error-error
    this.viewer.navigator.setAjaxHeaders({
      Authorization: this.apiRequestConfiguration.getToken!,
    });

    this.viewer.addHandler('zoom', e => {
      this.updateSizeReferenceBar(e.zoom);
    });

    this.viewer.addHandler('open', () => {
      this.viewer.viewport.goHome();

      const zoomFactor = 1;
      this.viewer.viewport.zoomBy(zoomFactor);

      // Add and apply border overlay to the image
      const imageBounds = this.viewer.world.getItemAt(0).getBounds();
      const borderOverlay = document.createElement('div');
      this.applyStyles(borderOverlay, this.getBorderOverlayStyle());
      this.viewer.addOverlay({
        element: borderOverlay,
        location: imageBounds,
      });
    });
  }

  private updateSizeReferenceBar(zoom: number): void {
    const viewportWithZoom =
      this.sampleImageMetadata!.size!.width!.pixel! / zoom;
    const containerWidth = this.viewer.container.getBoundingClientRect().width;

    const barSize =
      this.referenceSizeBarbreakpoints[this.refrenceBreakpointIndex] /
      this.sampleImageMetadata!.pixel!.width!.micro! /
      (viewportWithZoom / containerWidth);

    if (barSize > 125) {
      if (
        this.refrenceBreakpointIndex + 1 <
        this.referenceSizeBarbreakpoints.length
      ) {
        this.refrenceBreakpointIndex++;
      }
    } else if (barSize < 50) {
      if (this.refrenceBreakpointIndex > 0) {
        this.refrenceBreakpointIndex--;
      }
    }
    this.sampleSizeReference =
      this.referenceSizeBarbreakpoints[this.refrenceBreakpointIndex];
    this.referenceSizeBarWidth =
      this.referenceSizeBarbreakpoints[this.refrenceBreakpointIndex] /
      this.sampleImageMetadata!.pixel!.width!.micro! /
      (viewportWithZoom / containerWidth);
  }
```

**Comment by hrocha1** [comment]
@The-Aurora @iangilman I had the same issue with the initial load of dzi file. It seems like the `ajaxHeaders` are not correctly propagated from the Viewer if the source for the image is just a string containing the image URL. The URL gets transformed to

```
options = {
  tileSource: options
};
```

in the Viewer.open function and then the `addTiledImage` function ignores the Viewer global options and sets the `ajaxHeaders` to empty object

```
if (!$.isPlainObject(options.ajaxHeaders)) {
  options.ajaxHeaders = {};
}
```

instead of something like

```
if (!$.isPlainObject(options.ajaxHeaders)) {
  options.ajaxHeaders = this.ajaxHeaders;
}
```

Until this is fixed the possible workaround is to instantiate the viewer using full option objects for the `tileSources` and set the `ajaxHeaders` directly on them (although this might not be ideal if you need to update the headers over time).

Feels like this might be the same issue both the original author and @The-Aurora had, but I can create a new issue if needed.

**Comment by iangilman** [comment]
@hrocha1 Good catch!

I do think this is different enough from the original issue (which had to do specifically with the simple image tile source, not DZI), that it would be worth filing as a new issue.

It sounds like you've got a pretty good handle on it... would you be up for making a fix?

**Comment by hrocha1** [comment]
@iangilman ok, I will create a new issue for this. I'm already testing the fix as I need headers to work correctly for my use case (authorization via JWT). I will submit a pull request if I don't find any problems with it.

**Comment by iangilman** [comment]
@hrocha1 Wonderful!

**Comment by urapadmin** [comment]
I am using OSD with simple files and this original issue is a bit in the way of that (particularly having to supply the workaround with width and height prior to loading the image). 
Is anybody already working on this or shall I take a whack at the original issue?

**Comment by iangilman** [comment]
@urapadmin Apologies for the delayed response! Yes, if you're up for fixing it, that would be great!

**Comment by Aiosa** [comment]
Note that the behavior has changed significantly since this issue was raised, although the core logics still stays the same - getting image info involves actually downloading the whole image. 

What is worth considering is introducting a HTTP manager, instead of this dirty ajax headers being passed around. Because this makes stuff unmanageable. What do you do when the token updates regularly and you need to keep the fresh value always ready? How do you propagate the change to ajaxheaders without re-initializing the whole viewer?

In my viewer I've basically overriden the download tile start method to avoid this issue, and replaced the fetching logics with [a custom HTTPClient](https://github.com/RationAI/xopat/blob/release/v3/src/http-client.js) which consumes a secret store object. So instead of providing some hardcoded values via options, we can just provide some headers storage object for the client to consume, and build an API for headers and secret management.

**Comment by iangilman** [comment]
Interesting. Definitely worth considering finding a cleaner API for all that! It's just sort of grown organically...

**Comment by MatthiasDurivet** [comment]
Is this issue still open? I'm running into the same problem.

Update: fixed by providing the configuration inline instead of pointing to the .dzi file:
https://openseadragon.github.io/examples/tilesource-dzi/

**Comment by urapadmin** [comment]
I have solved this in my local fork a while ago but my pull request would only add to the organic growth of that part of the code. Not sure that's worth a pull request if a cleaner solution is the goal.

**Comment by iangilman** [comment]
I don't think we have really made much progress on coming up with a cleaner API. That said, there is this proposal from @Aiosa: https://github.com/openseadragon/openseadragon/pull/2882. Any thoughts there or here on how best to proceed are welcome!
